tests/python/contrib/test_csinn/infrastructure.py

```diff
@@ -0,0 +1,262 @@
+        config_file = os.path.join(location, file_name)
+        if not os.path.exists(config_file):
+            warnings.warn("Config file doesn't exist, resuming CSINN tests with default config.")
+            return
+        with open(config_file, mode="r") as config:
+            test_config = json.load(config)
+
+        cls.connection_type = test_config["connection_type"]
+        cls.host = test_config["host"]
+        cls.port = test_config["port"]
+        cls.target = test_config["target"]
+        cls.device_key = test_config.get("device_key") or ""
+        cls.cross_compile = test_config.get("cross_compile") or ""
+
+
+def get_cpu_op_count(mod):
+    """Traverse graph counting ops offloaded to TVM."""
+
+    class Counter(tvm.relay.ExprVisitor):
+        def __init__(self):
+            super().__init__()
+            self.count = 0
+
+        def visit_call(self, call):
+            if isinstance(call.op, tvm.ir.Op):
+                self.count += 1
+
+            super().visit_call(call)
+
+    c = Counter()
+    c.visit(mod["main"])
+    return c.count
+
+
+def skip_runtime_test():
+    """Skip test if it requires the runtime and it's not present."""
+    # CSINN codegen not present.
+    if not tvm.get_global_func("relay.ext.csinn", True):
+        print("Skip because CSINN codegen is not available.")
+        return True
+
+    # Remote device is in use or CSINN runtime not present
+    # Note: Ensure that the device config has been loaded before this check
+    if not Device.connection_type != "local" and not csinn.is_csinn_runtime_enabled():
+        print("Skip because runtime isn't present or a remote device isn't being used.")
+        return True
+
+
+def skip_codegen_test():
+    """Skip test if it requires the CSINN codegen and it's not present."""
+    if not tvm.get_global_func("relay.ext.csinn", True):
+        print("Skip because CSINN codegen is not available.")
+        return True
+
+
+def build_module(mod, target, params=None, enable_csinn=True, tvm_ops=0, csinn_partitions=1):
+    """Build module with option to build for CSINN."""
+    if isinstance(mod, tvm.relay.expr.Call):
+        mod = tvm.IRModule.from_expr(mod)
+    with tvm.transform.PassContext(opt_level=3, disabled_pass=["AlterOpLayout"]):
+        if enable_csinn:
+            mod = csinn.partition_for_csinn(mod, params)
+            tvm_op_count = get_cpu_op_count(mod)
+            assert tvm_op_count == tvm_ops, "Got {} TVM operators, expected {}".format(
+                tvm_op_count, tvm_ops
+            )
+            partition_count = 0
+            for global_var in mod.get_global_vars():
+                if "csinn" in global_var.name_hint:
+                    partition_count += 1
+
+            assert (
+                csinn_partitions == partition_count
+            ), "Got {} CSINN partitions, expected {}".format(partition_count, csinn_partitions)
+        relay.backend.te_compiler.get().clear()
```

Comment:
did you find this necessary?

Reply:
this part is used  to help check whether the annotation results are consistent with expectations